Lib/random.py

```diff
@@ -94,6 +94,18 @@ def __init__(self, x=None):
         self.seed(x)
         self.gauss_next = None
 
+    def __init_subclass__(cls, **kwargs):
+        # Only call self.getrandbits if the original random() builtin method
+        # has not been overridden or if a new getrandbits() was supplied.
+        if type(cls.__dict__.get('getrandbits')) is _FunctionType:
+            cls._randbelow = cls._randbelow_with_getrandbits
+        elif 'random' in cls.__dict__:
+            # There's an overridden random() method but no new getrandbits() method,
+            # so we can only use random() from here.
+            cls._randbelow = cls._randbelow_without_getrandbits
+        else:
+            cls._randbelow = getattr(cls, cls._randbelow.__name__)
```

Comment:
Is this needed?

Reply:
This is for the exotic case that a subclass overrides one of the private _randbelow implementations without providing random or getrandbits methods. Even though that would be considered a hack, inheritance mechanisms should still work and _randbelow_with_getrandbits or _randbelow_without_getrandbits get replaced. The test with Subclass5 is for this situation.

Reply:
It deserves a comment in the source though. I'll add it.

Reply:
`_randbelow` is an implementation detail. It isn't designed for overriding. The code will be simpler if don't support this hack.

Reply:
I still think that one extra line (I'm not really counting the else because if/elif without else looks questionable anyway) is worth it. Hack or not, our class-level manipulation of things should not make inheritance unpredictable.

Reply:
Serhiy is correct in pointing out the *_randbelow()* is an implementation detail.  We really don't want people using (overriding) this non public api.

Let's not feature-creep this optimization patch.  Ideally, it should be the smallest possible change that moves algorithmic decision to class instantiation time.    

Also, any possible "bug fix" to change the exception raised should be in a separate PR.  Unless we've guaranteed a particular exception (which is not usually the case), the C code is allowed to differ in small ways from the Python code.  That isn't to say that it shouldn't be changes, but that it is a separate discussion.

Reply:
ok: separate PR #6338 and corresponding https://bugs.python.org/issue33203 for the ValueError fix. Once that's settled I'll rebase this PR.

Speaking of not "feature-creeping" this simple optimization: would you prefer this patch then to retain the old behaviour that as soon as getrandbits was overridden once in a subclass, overriding random later again did not have an effect anymore on _randbelow (see Serhiy's Case 2 in https://bugs.python.org/msg314534 )?